src/lib.rs

```diff
@@ -6,18 +6,203 @@ An experimental x86_64 bootloader that works on both BIOS and UEFI systems.
+    }
+
+    /// Create a new, empty instance of DiskImageBuilder
+    pub fn empty() -> Self {
+        Self { files: Vec::new() }
+    }
+
+    /// Add or replace a kernel to be included in the final image.
+    pub fn set_kernel(&mut self, path: &Path) -> &mut Self {
+        self.add_or_replace_file(path, KERNEL_FILE_NAME)
+    }
+
+    /// Add or replace a ramdisk to be included in the final image.
+    pub fn set_ramdisk(&mut self, path: &Path) -> &mut Self {
+        self.add_or_replace_file(path, RAMDISK_FILE_NAME)
+    }
+
+    /// Configures the runtime behavior of the bootloader.
+    pub fn set_boot_config(&mut self, boot_config: &BootConfig) -> &mut Self {
+        let json =
+            serde_json::to_string_pretty(boot_config).expect("failed to serialize BootConfig");
+        let bytes = json.as_bytes();
+        self.add_or_replace_file_byte_content(bytes, CONFIG_FILE_NAME)
+    }
+
+    /// Add or replace a file from a byte array
+    pub fn add_or_replace_file_byte_content(&mut self, data: &[u8], target: &str) -> &mut Self {
+        let temp_path = temp_dir();
+        let file_name = temp_path.join("bytes.tmp");
+        fs::create_dir_all(temp_path).expect("Failed to create temp directory");
+        let mut temp_file =
+            fs::File::create(file_name.clone()).expect("Failed to create temp file");
+        temp_file
+            .write_all(data)
+            .expect("Failed to write data to temp file");
+        temp_file
+            .sync_all()
+            .expect("Failed to flush temp file to disk");
+
+        self.add_or_replace_file(&file_name, target)
+    }
+
+    /// Add or replace arbitrary files.
+    /// NOTE: You can overwrite internal files if you choose, such as EFI/BOOT/BOOTX64.EFI
+    /// This can be useful in situations where you want to generate an image, but not use the provided bootloader.
+    pub fn add_or_replace_file(&mut self, path: &Path, target: &str) -> &mut Self {
+        self.files.insert(
+            0,
+            DiskImageFile {
+                source: path.to_path_buf(),
+                destination: target.to_string(),
+            },
+        );
+        self
+    }
+    fn create_fat_filesystem_image(
+        &self,
+        internal_files: BTreeMap<&str, &Path>,
+    ) -> anyhow::Result<NamedTempFile> {
+        let mut local_map = BTreeMap::new();
+
+        for k in internal_files {
+            local_map.insert(k.0, k.1);
+        }
+
+        for f in self.files.as_slice() {
+            local_map.insert(&f.destination, f.source.as_path());
+        }
+
+        let out_file = NamedTempFile::new().context("failed to create temp file")?;
+        fat::create_fat_filesystem(local_map, out_file.path())
+            .context("failed to create BIOS FAT filesystem")?;
+
+        Ok(out_file)
+    }
+    #[cfg(feature = "bios")]
+    /// Create an MBR disk image for booting on BIOS systems.
+    pub fn create_bios_image(&self, image_filename: &Path) -> anyhow::Result<()> {
```

Comment:
I don't like the name `image_filename` because it's a full path, not just a file name. Maybe somethingl like `out_path`/`target_path`/`image_path`?

Reply:
Sure!